python/python/bystro/supervised_ppca/gf_generative_pt.py

```diff
@@ -0,0 +1,715 @@
+
+    def _initialize_variables(self, X):
+        """
+        Initializes the variables of the model
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,p)
+            The data
+
+        Returns
+        -------
+        W_ : tf.Variable-like,(n_components,p)
+            The loadings of our latent factor model
+
+        sigmal_ : list
+            A list of the isotropic noises for each group
+        """
+        model = dp.PCA(self.n_components)
+        S_hat = model.fit_transform(X)
+        W_init = model.components_.astype(np.float32)
+        W_ = torch.tensor(W_init, requires_grad=True)
+        X_recon = np.dot(S_hat, W_init)
+        diff = np.mean((X - X_recon) ** 2)
+        sinv = softplus_inverse_np(diff * np.ones(1).astype(np.float32))
+        sigmal_ = [
+            torch.tensor(sinv, requires_grad=True) for i in range(self.n_groups)
+        ]
+        return W_, sigmal_
+
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of variables learned by pytorch
+
+        Sets
+        ----
+        W_ : np.array-like,(n_components,p)
+            The loadings
+
+        sigmas_ : np.array-like,(p,)
+            The diagonal variances
+        """
+        self.W_ = trainable_variables[0].detach().numpy()
+        self.sigmas_ = np.zeros(self.p)
+        for k in range(self.n_groups):
+            sigma2 = nn.Softplus()(trainable_variables[k + 1])
+            self.sigmas_[self.groups == k] = sigma2
+
+    def _test_inputs(self, X, groups):
+        """
+        Just tests to make sure data is numpy array
+        """
+        if isinstance(X, np.ndarray) is False:
+            raise ValueError("X is numpy array")
+        if isinstance(groups, np.ndarray) is False:
+            raise ValueError("groups is numpy array")
+        if X.shape[1] != len(groups):
+            raise ValueError("Dimensions do not match")
+
+
+class FactorAnalysispt(BasePCASGDModel):
+    def __init__(
+        self, n_components=2, prior_options=None, training_options=None
+    ):
+        """
+        This implements factor analysis which allows for each covariate to
+        have it's own isotropic noise. No analytic solution that I know of
+        but fortunately with SGD it doesn't matter.
+
+        Parameters
+        ----------
+        n_components : int,default=2
+            The latent dimensionality
+
+        training_options : dict,default={}
+            The options for gradient descent
+
+        prior_options : dict,default={}
+            The options for priors on model parameters
+        """
+        super().__init__(
+            n_components=n_components,
+            prior_options=prior_options,
+            training_options=training_options,
+        )
+
+    def __repr__(self):
+        out_str = "FactorAnalysispt object\n"
```

Comment:
same comment as before about pythonic reprs.

I think from here to the end of the file all of the remaining concerns have been touched on previously

Reply:
Yeah changed all the reprs.